algebird-core/src/main/scala/com/twitter/algebird/macros/MonoidMacro.scala

```diff
@@ -0,0 +1,37 @@
+package com.twitter.algebird.macros
+
+import scala.language.experimental.macros
+import scala.reflect.macros.Context
+import scala.reflect.runtime.universe._
+
+import com.twitter.algebird._
+
+object MonoidMacro {
+  def caseClassMonoid[T](c: Context)(implicit T: c.WeakTypeTag[T]): c.Expr[Monoid[T]] = {
+    import c.universe._
+
+    val tpe = weakTypeOf[T]
+
+    val isCaseClass = tpe.typeSymbol.isClass && tpe.typeSymbol.asClass.isCaseClass
+    if (!isCaseClass)
+      c.abort(c.enclosingPosition, s"$T is not a clase class")
+
+    val params = tpe.declarations.collect {
+      case m: MethodSymbol if m.isCaseAccessor => m
+    }.toList
+
+    val plusList = params.map(param => q"implicitly[_root_.com.twitter.algebird.Monoid[$param]].plus(l.$param, r.$param)")
+    val zerosList = params.map(param => q"implicitly[_root_.com.twitter.algebird.Monoid[$param]].zero")
+
+    val companion = tpe.typeSymbol.companionSymbol
+
+    val res = q"""
+    new _root_.com.twitter.algebird.Monoid[$T] {
+      def plus(l: $T, r: $T): $T = $companion.apply(..$plusList)
+      def zero: $T = $companion.apply(..$zerosList)
```

Comment:
One way to make `isNonZero` more efficient is changing `def zero` to `val zero`. This will ensure that zero is only computed once. Then any call to `isNonZero(x)` will just return `x != zero`. This should be reasonably efficient.

What did you have in mind for an implementation of `isNonZero`?